crates/edr_solidity/src/bytecode_trie.rs

```diff
@@ -1,83 +1,417 @@
+                    });
+                }
+            } else if let Some(node) = cursor.child_nodes.get(&key[index]) {
                 cursor = node;
+            } else if !cursor.is_root() {
+                return Some(TrieSearch::LongestPrefixNode {
+                    node: cursor,
+                    diff_index: index,
+                    match_: cursor.match_.clone(),
+                });
+            }
+
+            index += 1;
+        }
+
+        cursor.match_.as_ref().and_then(|item| {
+            if cursor.prefix.range_end == key.len() {
+                // If the cursor's range ends where the key ends, we have a hit.
+                Some(TrieSearch::ExactHit(item.clone()))
             } else {
-                return Some(TrieSearch::LongestPrefixNode(cursor));
+                // Otherwise the cursor's range is greater than the key's length which means the
+                // key is a prefix of the match.
+                None
             }
+        })
+    }
+
+    /// Whether the node is a root node.
+    pub fn is_root(&self) -> bool {
+        matches!(self.prefix.key, TrieKey::Root)
+    }
+
+    fn new_leaf(new_item: T) -> Self {
+        Self {
+            child_nodes: HashMap::default(),
+            // We have to include leaf nodes as descendants of themselves, because
+            // leaf nodes can be returned in `TrieSearch::LongestPrefixNode` results after
+            // eliminating one-way branching.
+            descendants: vec![new_item.clone()],
+            match_: Some(new_item.clone()),
+            prefix: TriePrefix {
+                range_end: new_item.key().len(),
+                key: TrieKey::Key(new_item),
+            },
         }
+    }
+
+    /// Create a new split node with the split prefix.
+    /// Fill it by calling `Self::fill_split_node` on it after swapping it with
+    /// the node to split.
+    fn new_split_node(split_index: usize, new_item: T) -> Self {
+        BytecodeTrie {
+            prefix: TriePrefix {
+                range_end: split_index,
+                key: TrieKey::Key(new_item),
+            },
+
+            // These be filled in by calling `Self::fill_split_node`
+            child_nodes: HashMap::default(),
+            descendants: Vec::default(),
+            match_: None,
+        }
+    }
+
+    /// Fill the split node from the node with the node
+    /// and the new key as children. If the new key is shorter than the
+    /// node's prefix, the new split node will be a match for the new
+    /// key. Panics if `split_index > new_key.key().len()`
+    fn fill_split_node(&mut self, node_to_split: BytecodeTrie<T>, new_item: T) {
+        // We use the descendants to keep track of insertion order, so it's
+        // important to preserve that order here
+        let mut descendants = Vec::with_capacity(node_to_split.descendants.len() + 1);
+        descendants.extend(node_to_split.descendants.iter().cloned());
+        descendants.push(new_item.clone());
```

Comment:
I don't think this variable is used in this function

Reply:
Uhh that's something I broke when fixing the last bug. Thanks for catching it! Fixed + tested in 
[bc378e0](https://github.com/NomicFoundation/edr/pull/775/commits/bc378e065ff3c49e9a69f6c1bbedff4ae690ebca)